File: python/src/project_euler/problem004.py
# A palindromic number reads the same both ways. The largest palindrome made
# from the product of two 2-digit numbers is 9009 = 91 × 99.
#
# Find the largest palindrome made from the product of two 3-digit numbers.

def main():
    largest_so_far = 0
    for i in range(100,1000):
        for j in range(100, 1000):
            if is_palindrome(i*j) and i*j > largest_so_far:
                largest_so_far = i*j
    return largest_so_far

def is_palindrome(n):
    s = str(n)
    return s == s[::-1]

File: typescript/src/problems/problem_004.ts

// Problem 4:
// A palindromic number reads the same both ways.
// The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.
// Find the largest palindrome made from the product of two 3-digit numbers.

export function main(): number {

    function isPalindrome(n: number): boolean {
        const N = n.toString()
        const N_reverse = N.split("").reverse().join("")
        return N == N_reverse
    }

    let largestSoFar = 0
    for(let i = 1; i < 1000; i++) {
        for(let j = 1; j <= i; j++) { // only test the lower triangular region of the iXj matrix, where j <= i
            if(isPalindrome(i*j) && i*j >= largestSoFar) {
                largestSoFar = i*j
            }
        }
    }
    return largestSoFar
}
